src/main/java/com/google/api/gax/grpc/PageImpl.java

```diff
@@ -0,0 +1,166 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.api.gax.grpc;
+
+import com.google.api.gax.core.Page;
+import com.google.api.gax.protobuf.ValidationException;
+import com.google.common.base.Throwables;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.Iterators;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+
+import io.grpc.StatusRuntimeException;
+
+import java.util.Iterator;
+import java.util.concurrent.Future;
+
+class PageImpl<RequestT, ResponseT, ResourceT> implements Page<RequestT, ResponseT, ResourceT> {
+
+  private final FutureCallable<RequestT, ResponseT> callable;
+  private final PageStreamingDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor;
+  private final CallContext<RequestT> context;
+  private ResponseT response;
+
+  /** */
+  public PageImpl(
+      FutureCallable<RequestT, ResponseT> callable,
+      PageStreamingDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor,
+      CallContext<RequestT> context) {
+    this.callable = callable;
+    this.pageDescriptor = pageDescriptor;
+    this.context = context;
+
+    // Make the API call eagerly
+    this.response = getUnchecked(callable.futureCall(context));
+  }
+
+  @Override
+  public Iterator<ResourceT> iterator() {
+    return pageDescriptor.extractResources(response).iterator();
+  }
+
+  @Override
+  public boolean hasNextPage() {
+    return !getNextPageToken().equals(pageDescriptor.emptyToken());
+  }
+
+  @Override
+  public Object getNextPageToken() {
+    return pageDescriptor.extractNextToken(response);
+  }
+
+  @Override
+  public Page<RequestT, ResponseT, ResourceT> getNextPage() {
+    if (!hasNextPage()) {
+      throw new ValidationException(
+          "Could not complete getNextPage operation: there are no more pages to retrieve.");
```

Comment:
I think we should use NoSuchElementException, like Iterator uses for next():

https://docs.oracle.com/javase/7/docs/api/java/util/Iterator.html

Maybe we should also consider a naming scheme like "nextPage()". (lower priority)